core/src/replay_stage.rs

```diff
@@ -1804,10 +1912,37 @@ impl ReplayStage {
                 );
             }
         }
+
+        // send accumulated excute-timings to service thread to update cost model
+        cost_update_sender
+            .send(execute_timings)
+            .expect("send execution cost update to cost_model");
+
         inc_new_counter_info!("replay_stage-replay_transactions", tx_count);
         did_complete_bank
     }
 
+    fn persist_cost_table(blockstore: &Blockstore, cost_model: &RwLock<CostModel>) {
+        let cost_model_read = cost_model.read().unwrap();
+        let cost_table = cost_model_read.get_instruction_cost_table();
+        let db_records = blockstore.read_program_costs().expect("read programs");
+
+        // delete records from blockstore if they are no longer in cost_table
+        db_records.iter().for_each(|(pubkey, _)| {
+            if !cost_table.iter().any(|(key, _)| key == pubkey) {
```

Comment:
I think you could simplify this by using `HashMap::get()`:
```
if cost_table.get(&pubkey).is_none() {
   ...
```

I think that is a little easier to read, but also, the ordering on `cost_table.iter()` [is arbitrary](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.iter). As such, I think using `.any()` could incur the cost of running through the entire iterator in the worst case scenario (where everything in `db_records` is no longer in `cost_table`) for each element in `db_records`

Reply:
Nice! thanks for explaining it. Makes good sense. I don't know where my twisted statement came from, but your suggestion definitely makes it much better.